decentralized-message-queue/src/DMQ/Protocol/LocalMsgNotification/Type.hs

```diff
@@ -0,0 +1,181 @@
+{-# LANGUAGE DeriveFoldable           #-}
+{-# LANGUAGE FlexibleInstances        #-}
+{-# LANGUAGE PolyKinds                #-}
+{-# LANGUAGE StandaloneDeriving       #-}
+{-# LANGUAGE StandaloneKindSignatures #-}
+{-# LANGUAGE TypeData                 #-}
+{-# LANGUAGE TypeFamilies             #-}
+
+-- | Defines types for the local message notification protocol
+--
+module DMQ.Protocol.LocalMsgNotification.Type where
+
+import Control.DeepSeq
+import Data.Kind
+import Data.List.NonEmpty
+import Network.TypedProtocol.Core
+import Ouroboros.Network.Util.ShowProxy
+
+-- | There are some constraints of the protocol that are not captured in the
+-- types of the messages, but are documented with the messages. Violation
+-- of these constraints is also a protocol error. The constraints are intended
+-- to ensure that implementations are able to work in bounded space.
+--
+instance Protocol (LocalMsgNotification msg) where
+  -- | The messages in the protocol.
+  --
+  -- There are two ways to ask for messages, blocking and
+  -- non-blocking
+  -- TODO: blocking only should suffice?.
+  -- They otherwise have the same semantics.
+  --
+  data Message (LocalMsgNotification msg) from to where
+
+    -- | Request a list of messages from the server,
+    --
+    -- * The blocking case is used when the server has announced it has no
+    --   further messages to provide, and otherwise the non-blocking request
+    --   must be used.
+    --
+    -- TODO: this request should have the max number of messages that shall
+    -- be returned?
+    MsgRequest
+      :: forall blocking msg.
+         SingBlockingStyle blocking
+      -> Message (LocalMsgNotification msg) StIdle (StBusy blocking)
+
+    -- | Reply with a list of messages, along with a flag indicating
+    -- whether the server has more messages that it can provide.
+    --
+    -- TODO: maybe this should just be always a blocking request?
+    --
+    MsgReply
+      :: forall blocking msg.
+         BlockingReplyList blocking msg
+      -> HasMore
+      -> Message (LocalMsgNotification msg) (StBusy blocking) StIdle
+
+    -- | Termination message, initiated by the server when the client is
+    -- making a blocking call for more messages.
+    --
+    -- TODO: is this message desireable/necessary?
+    --
+    MsgServerDone
+      :: Message (LocalMsgNotification msg) (StBusy StBlocking) StDone
+
+    -- | The client can terminate the exchange when it has the agency
+    --
+    MsgClientDone
+      :: Message (LocalMsgNotification msg) StIdle StDone
+
+  type StateAgency  (StBusy blocking)  = ServerAgency
+  type StateAgency  StIdle  = ClientAgency
+  type StateAgency  StDone  = NobodyAgency
+
+  type StateToken = SingMsgNotification
+
+instance ( NFData msg
+         ) => NFData (Message (LocalMsgNotification msg) from to) where
+  rnf (MsgRequest block)  = rnf block
+  rnf (MsgReply a !_more) = rnf a
+  rnf MsgServerDone{}     = ()
+  rnf MsgClientDone{}     = ()
```

Comment:
We never needed `NFData` instances for `Messages`, do we need them? If not I'll remove them; it's one thing less to maintain.